### ISSUE TITLE
ARM based docker image

### ISSUE THREAD

**Issue by vkuznecovas** [apache/pulsar#12944]
**Is your enhancement request related to a problem? Please describe.**
I'd like to see an ARM based docker image for pulsar. The problem is that under new macs the pulsar containers have to be run under emulation which makes them horribly slow.

**Describe the solution you'd like**
ARM based docker image should be available in the apachepulsar/pulsar repo. 

**Describe alternatives you've considered**
Building a native image has not proved trivial, so ideally someone with more knowledge of the project would be able to create one.

**Comment by michaeljmarshall** [comment]
+1. This would be a good feature for end users. My only concern as a contributor is that it will increase build times. That shouldn't stop this feature, though.

Ideally, we'll be able to follow the instructions here: https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/. I haven't done a mutli-arch build using maven before (we build our docker images using maven), so that is one nuance.

**Comment by kezhenxu94** [comment]
I'm glad to work on this if there is nobody already working on it?

**Comment by krissetto** [comment]
I am also interested in support for ARM docker images, especially for testing locally on Apple Silicon based machines

**Comment by kezhenxu94** [comment]
> I am also interested in support for ARM docker images, especially for testing locally on Apple Silicon based machines

Thank you @krissetto !! Will let you know when I have a docker image for testing on ARM!

**Comment by lorthirk** [comment]
+1!

**Comment by jibbolo** [comment]
+1

**Comment by kezhenxu94** [comment]
Hi all in this thread, I just pushed a multi-arch image to my own repo, anyone interested is welcome to pull and test it in your environment. https://hub.docker.com/repository/docker/kezhenxu94/pulsar/tags?page=1&ordering=last_updated

To test, just replace the docker image with `kezhenxu94/pulsar`, example:

```
docker run -it \                                                                                                                  
  -p 6650:6650 \
  -p 8080:8080 \
  --mount source=pulsardata,target=/pulsar/data \
  --mount source=pulsarconf,target=/pulsar/conf \
  kezhenxu94/pulsar \
  bin/pulsar standalone
```

Note this is not official and not well tested now, I'll test, polish and then open a PR so everyone can review on the build process.

**Comment by vkuznecovas** [comment]
> Hi all in this thread, I just pushed a multi-arch image to my own repo, anyone interested is welcome to pull and test it in your environment. https://hub.docker.com/repository/docker/kezhenxu94/pulsar/tags?page=1&ordering=last_updated
> 
> To test, just replace the docker image with `kezhenxu94/pulsar`, example:
> 
> ```
> docker run -it \                                                                                                                  
>   -p 6650:6650 \
>   -p 8080:8080 \
>   --mount source=pulsardata,target=/pulsar/data \
>   --mount source=pulsarconf,target=/pulsar/conf \
>   kezhenxu94/pulsar \
>   bin/pulsar standalone
> ```
> 
> Note this is not official and not well tested now, I'll test, polish and then open a PR so everyone can review on the build process.
> 
> Any feedback are also welcome and appreciated 🙇

I've tested it out, seems to work great. Thanks!

**Comment by krissetto** [comment]
I have also started testing it on an M1 device, and it seems to work 😄 I haven't gone very deep into it yet though, so i'll be back with feedback if I encounter any specific issues with it. Thanks for the build! Ping when you have a PR too since i'm curious

**Comment by kezhenxu94** [comment]
> I have also started testing it on an M1 device, and it seems to work 😄 I haven't gone very deep into it yet though, so i'll be back with feedback if I encounter any specific issues with it. Thanks for the build! Ping when you have a PR too since i'm curious

Sure, will keep you all in this thread posted. Thanks all for testing. 🙇

**Comment by tdrozdowski** [comment]
I can also validate that this appears to work on M1 Pro Max based machine.  Would love to see the PR so we can get an official image; this works for local development in the meantime (and eliminates yet another tie to an Intel based machine!).  Thanks so much!

**Comment by kezhenxu94** [comment]
Heads up: The last (hard) step is to build the pulsar client with multiple architectures. Everything else is ready. I have some ideas to build pulsar client in multi architectures but need some time to verify, which may be also ready in 2 weeks (due to public holidays and my personal stuffs)

**Comment by kezhenxu94** [comment]
> @kezhenxu94 thanks for working on this, our company relies on pulsar and a recent move to Mac M1 has led us to run into a ton of problems with local development. Would it be possible to share the Dockerfile so that we could temporarily build arm supported images for older versions of pulsar? If not, do you have a timeline so we can be ready to start switching out our images?

I can draft a PR of what I've done tomorrow (in my time zone) so that you can try it out. Stay tune please.

**Comment by kezhenxu94** [comment]
Hi, I've just drafted a PR https://github.com/apache/pulsar/pull/14005

**Comment by krissetto** [comment]
Are there any updates on this issue? I see the draft PR hasn't changed since last month's revision

**Comment by heesung-sn** [comment]
For those of you who are stuck at this issue, I was able to create ARM docker images on my M1 Mac by the following simple change. The test dockers from the following builds passed the shade integration test on my laptop.

```
# docker builds
mvn -B -f docker/pulsar/pom.xml install -am -Pdocker,-main -DskipTests -Ddocker.nocache=true
mvn -B -f docker/pulsar-all/pom.xml install -am -Pdocker,-main -DskipTests -Ddocker.nocache=true
mvn -B -f tests/docker-images/pom.xml install -am -Pdocker,-main -DskipTests

# shade test run
mvn -B -ntp -DredirectTestOutputToFile=false -f tests/pom.xml test -DShadeTests -DtestForkCount=1 -DtestReuseFork=false
```

```
hsohn@HeesungSohns-MacBook-Pro pulsar % git diff
diff --git a/build/docker/Dockerfile b/build/docker/Dockerfile
index 4b6b8a3a84..da4b04e9f0 100644
--- a/build/docker/Dockerfile
+++ b/build/docker/Dockerfile
@@ -51,8 +51,13 @@ RUN cd /usr/src/gmock && cmake . && make && cp libgmock.a /usr/lib
 # Include gtest parallel to speed up unit tests
 RUN git clone https://github.com/google/gtest-parallel.git

-ENV JAVA_HOME=/usr/lib/jvm/java-1.8.0-openjdk-amd64
-ENV JAVA_HOME_11=/usr/lib/jvm/java-1.11.0-openjdk-amd64
+RUN java_home=$(ls -d1 /usr/lib/jvm/java-1.8.0-openjdk-*) ; \
+    ln -s "$java_home" /usr/lib/jvm/java8 ; \
+    java_home_11=$(ls -d1 /usr/lib/jvm/java-1.11.0-openjdk-*) ; \
+    ln -s "$java_home_11" /usr/lib/jvm/java11
+
+ENV JAVA_HOME /usr/lib/jvm/java8
+ENV JAVA_HOME_11 /usr/lib/jvm/java11

 ## Website build dependencies

diff --git a/docker/pulsar/Dockerfile b/docker/pulsar/Dockerfile
index 711b9509d3..1bfb622574 100644
--- a/docker/pulsar/Dockerfile
+++ b/docker/pulsar/Dockerfile
@@ -71,9 +71,12 @@ RUN update-alternatives --install /usr/bin/python python /usr/bin/python3 10
 # 3. /pulsar/download - functions write to this directory
 # 4. /pulsar - hadoop writes to this directory
 RUN mkdir /pulsar && chmod g+w /pulsar
+RUN java_home=$(ls -d1 /usr/lib/jvm/java-11-openjdk-*) ; \
+    ln -s "$java_home" /usr/lib/jvm/java
+
+ENV JAVA_HOME /usr/lib/jvm/java
+RUN echo networkaddress.cache.ttl=1 >> /usr/lib/jvm/java/conf/security/java.security

-ENV JAVA_HOME /usr/lib/jvm/java-11-openjdk-amd64
-RUN echo networkaddress.cache.ttl=1 >> /usr/lib/jvm/java-11-openjdk-amd64/conf/security/java.security
 ADD target/python-client/ /pulsar/pulsar-client

 ENV PULSAR_ROOT_LOGGER=INFO,CONSOLE
diff --git a/docker/pulsar/pom.xml b/docker/pulsar/pom.xml
index 048b3da7cb..d4aa4d2eb1 100644
--- a/docker/pulsar/pom.xml
+++ b/docker/pulsar/pom.xml
@@ -75,7 +75,7 @@
                   <executable>${project.basedir}/../../pulsar-client-cpp/docker/build-wheels.sh</executable>
                   <arguments>
                     <!-- build python 3.8 -->
-                    <argument>3.8 cp38-cp38 manylinux2014 x86_64</argument>
+                    <argument>3.8 cp38-cp38 manylinux2014 aarch64</argument>
                   </arguments>
                 </configuration>
               </execution>
diff --git a/docker/pulsar/scripts/install-pulsar-client.sh b/docker/pulsar/scripts/install-pulsar-client.sh
index 56702b10db..adba8ed1cd 100755
--- a/docker/pulsar/scripts/install-pulsar-client.sh
+++ b/docker/pulsar/scripts/install-pulsar-client.sh
@@ -22,4 +22,11 @@ set -x

 PYTHON_MAJOR_MINOR=$(python3 -V | sed -E 's/.* ([[:digit:]]+)\.([[:digit:]]+).*/\1\2/')
 WHEEL_FILE=$(ls /pulsar/pulsar-client | grep "cp${PYTHON_MAJOR_MINOR}")
+
+# The following `update build-essential python3-dev` are added due to grpcio `cc` not found error
+# TODO: remove these lines if the error got resolved.
+# WARNING: currently grpcio build takes long.
+apt-get update
+apt-get -y install build-essential python3-dev
+
 pip3 install /pulsar/pulsar-client/${WHEEL_FILE}[all]
diff --git a/tests/docker-images/java-test-image/Dockerfile b/tests/docker-images/java-test-image/Dockerfile
index 564182534a..ca7c719cfb 100644
--- a/tests/docker-images/java-test-image/Dockerfile
+++ b/tests/docker-images/java-test-image/Dockerfile
@@ -40,8 +40,11 @@ RUN sed -i "s|http://archive\.ubuntu\.com/ubuntu/|${UBUNTU_MIRROR:-mirror://mirr
      && apt-get -y dist-upgrade \
      && apt-get -y install openjdk-11-jdk-headless

-ENV JAVA_HOME /usr/lib/jvm/java-11-openjdk-amd64
-RUN echo networkaddress.cache.ttl=1 >> /usr/lib/jvm/java-11-openjdk-amd64/conf/security/java.security
+RUN java_home=$(ls -d1 /usr/lib/jvm/java-11-openjdk-*) ; \
+    ln -s "$java_home" /usr/lib/jvm/java
+ENV JAVA_HOME /usr/lib/jvm/java
+RUN echo networkaddress.cache.ttl=1 >> /usr/lib/jvm/java/conf/security/java.security
+

 # /pulsar/bin/watch-znode.py requires python3-kazoo
 # /pulsar/bin/pulsar-managed-ledger-admin requires python3-protobuf
diff --git a/tests/docker-images/latest-version-image/Dockerfile b/tests/docker-images/latest-version-image/Dockerf
ile
index f5c6726831..e9a2ba1c0d 100644
--- a/tests/docker-images/latest-version-image/Dockerfile
+++ b/tests/docker-images/latest-version-image/Dockerfile
@@ -28,8 +28,8 @@ RUN rm -rf /var/lib/apt/lists/* && apt-get update
 RUN apt-get install -y procps curl git build-essential

 ENV GOLANG_VERSION 1.15.8
-
-RUN curl -sSL https://golang.org/dl/go$GOLANG_VERSION.linux-amd64.tar.gz \
+RUN export arch=$(uname -m | sed -r 's/aarch64/arm64/g' |  awk '!/arm64/{$0="amd64"}1'); \
+curl -sSL https://golang.org/dl/go$GOLANG_VERSION.linux-$arch.tar.gz \
                | tar -C /usr/local -xz

 # RUN wget https://dl.google.com/go/go1.13.3.linux-amd64.tar.gz && tar -xvf go1.13.3.linux-amd64.tar.gz && mv go /
usr/local
```

**Comment by rhenwood-arm** [comment]
Hi All, I'm interested to learn if I can support this effort.

**Comment by dave2wave** [comment]
> Hi All, I'm interested to learn if I can support this effort.

PRs are welcome!

**Comment by grishick** [comment]
Looks like a related PR has been closed, but ARM image has not yet been published. I would appreciate any update on this

**Comment by michaeljmarshall** [comment]
Now that https://github.com/apache/pulsar/pull/15142 is merged, there are no technical blockers for this work to continue on in the Pulsar project. I'll follow up on this one to try to push this through for the 2.11 release.

**Comment by michaeljmarshall** [comment]
New mailing list thread: https://lists.apache.org/thread/52bncdx2j4j982m9xob85sjmsr23tgzt

**Comment by tisonkun** [comment]
This can be related to #16652 .

Also, when I build on Apple M1 (ARM64), I find that `pulsar-client-cpp/docker/manylinux1/Dockerfile` binds to x86_64 and thus the build will fail with:

```
ERROR: pulsar_client-2.11.0-cp38-cp38-manylinux1_x86_64.whl is not a supported wheel on this platform.
```

**Comment by tisonkun** [comment]
Trying to workaround with:

```
$ pulsar-client-cpp/docker/build-wheels.sh 3.8 cp38-cp38 manylinux2014 aarch64
$ mvn package -Pdocker,-main -am -pl docker/pulsar-all -DskipTests -DskipBuildPythonClient=true -DskipSourceReleaseAssembly=true
```

We should integrate this snippet into build scripts, though.

FAILED with Python client depends on grpcio = 1.27.2, which doesn't support aarch64. A newer version can help, see also https://github.com/grpc/grpc/pull/25418 and https://github.com/grpc/grpc/issues/21283.

cc @BewareMyPower @shibd @RobertIndie

**Comment by tisonkun** [comment]
It seems all the issues are included in https://github.com/apache/pulsar/issues/12944#issuecomment-1103444708. I'll try it out.

**Comment by tisonkun** [comment]
With #17733 we should be able to build ARM-based docker image manually. Integrating it with automation or CI is another topic.

**Comment by eolivelli** [comment]
if you want to run pulsar standalone without docker you can simply add this to conf/standalone.conf, this will disable RocksDB

`ledgerStorageClass=org.apache.bookkeeper.bookie.SortedLedgerStorage`

**Comment by merlimat** [comment]
Since Pulsar 2.11.0, there is no problem in running on ARM64 with default configs

**Comment by zhongduo** [comment]
Any chance to support ARM64 docker in dockerhub?

**Comment by merlimat** [comment]
As of now, you can build ARM64 image, though we don't yet have multi-arch image. It needs some work on the maven-docker plugin to do so

**Comment by onobc** [comment]
> But I have this error: WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested

I see that warning too, but other than that it works like a charm for me.

**Comment by merlimat** [comment]
Yes, this is going to be available in 3.0.0 which is being released this week.
